### PR TITLE
Guard against negative multiplicities in TopK

### PR DESCRIPTION
**Pull request by tokenrove**
~Opening a draft to get feedback on two alternative approaches included as WIP commits over the inner function approach used in #17918.~  This is now adjusted to use the `MaybeValidatingRow` abstraction in two places.  I still want to think about this and look at this code again with fresh eyes and decide if I like it better than the alternative, but I'm opening this up for review to get some feedback.  I can always drop this back to the earlier implementation (https://github.com/tokenrove/materialize/commit/8289d0fe92231a2a10ad7a2f1b8318e4561752dd) if we decide this is horrible.

I still haven't been able to convince myself that we want to do this, but it is clear that these reductions like TopK are the only places we can legitimately catch this cases, and that in the absence of checks like this, strange inconsistencies can occur -- even if they're only the consequence of bugs elsewhere, we can only effectively detect them here.  So largely I think this is the right thing to do.

### Motivation

 * This PR fixes a recognized bug: #17963.

### Tips for reviewer

Probably worth reviewing the commits individually, as the first adjusts some whitespace.

The `MaybeValidatingRow` trait seems to clean up the code, but I myself need to step away from it for a bit to come back and see if that's really true, since I've been staring at it a lot lately.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/compute/src/render/top_k.rs
Comment (teskje) [inline src/compute/src/render/top_k.rs:392]: You could be more aggressive with avoiding rightward-drift by `continue`ing on the diff checks. I'd personally prefer this but I know different people think differently, so I'll leave that up to you and @vmarcos.

File: src/compute/src/render/mod.rs
Comment (teskje) [inline src/compute/src/render/mod.rs:928]: Nit: It would be appropriate to `use` these traits directly, rather than their modules. We do so throughout the codebase and their names are unambiguous.

File: src/compute/src/render/reduce.rs
Comment (teskje) [inline src/compute/src/render/reduce.rs:781]: This changes how the error handling behaves. Previously we reported all invalid counts we found, the new version only reports the first instance. Is that intended?

File: src/compute/src/render/mod.rs
Comment (teskje) [inline src/compute/src/render/mod.rs:926]: Not sure if that improves things, but you could encode this contract in the type system by having, e.g. an `into_error` trait method that returns an `Option<&dyn Fn(E) -> Self>` (or whatever actually type-checks). Then non-validating types could return `None` here, so there is no error constructor that could be invoked for them.

Maybe that makes things more complicated than they need to be...

**Review by teskje — APPROVED** [review]
LGTM, pending Marcos' approval. I find the `MaybeValidatingRow` slightly more readable I think, though it's unfortunately not a huge improvement.

File: src/compute/src/render/reduce.rs
Comment (vmarcos) [inline src/compute/src/render/reduce.rs:779]: I think this last line of the comment should probably be removed?

**Review by vmarcos — COMMENTED** [review]
Overall, this looks good to me. I have one specific question regarding making the top-k stage of monotonic top-k validating here or employing an `ensure_monotonic` on the input as in #14607. We should probably agree on which alternative would be preferable prior to merging.

**Comment by vmarcos** [comment]
Post-review comment: Should we also add a test based on the scenario in https://github.com/MaterializeInc/materialize/issues/17963#issuecomment-1477630193 to confirm that we produce errors adequately? This can be done as an additional `mzcompose` workflow in `test/cluster`.

**Comment by tokenrove** [comment]
> Post-review comment: Should we also add a test based on the scenario in [#17963 (comment)](https://github.com/MaterializeInc/materialize/issues/17963#issuecomment-1477630193) to confirm that we produce errors adequately? This can be done as an additional `mzcompose` workflow in `test/cluster`.

Yes, I think each of these kind of similar issues should all have tests, since they're easy to reproduce in unsafe mode.  I was thinking that I would do these as testdrive tests, just assuring we get the user-level error back, since we already have a test that assures errors go into the logs, and testdrive should be a little lighter-weight.  (Though perhaps testdrive should also grow a function to inspect logs.)

File: src/compute/src/render/top_k.rs
Comment (tokenrove) [reply src/compute/src/render/top_k.rs:392]: Good call; this is my preference too, and it would be nice to renovate this code while I'm touching it.

File: src/compute/src/render/mod.rs
Comment (tokenrove) [reply src/compute/src/render/mod.rs:926]: That's an interesting idea.  I'll play with it a bit.

File: src/compute/src/render/mod.rs
Comment (tokenrove) [reply src/compute/src/render/mod.rs:1]: My instinct is that `mod.rs` is usually the right place for types used throughout the module, but especially if we think we'll grow more validation code within render, I'd be happy to move it to `validation.rs`.

File: src/compute/src/render/top_k.rs
Comment (tokenrove) [reply src/compute/src/render/top_k.rs:139]: Cool, I'll set validating to false in this call, then.

File: src/compute/src/render/top_k.rs
Comment (tokenrove) [reply src/compute/src/render/top_k.rs:405]: I think this is one of the main things to evaluate about this approach.  I think it keeps the code cleaner, but at this additional cost (plus the trickiness of the errors also going through this negation).  How about this permutation, which is I think a little less clear but not too much so?

```diff
diff --git a/src/compute/src/render/top_k.rs b/src/compute/src/render/top_k.rs
index daa03a2f3..37b45397d 100644
--- a/src/compute/src/render/top_k.rs
+++ b/src/compute/src/render/top_k.rs
@@ -264,10 +266,12 @@ where
             G: Scope,
             G::Timestamp: Lattice,
         {
-            if validating {
+            let input = collection.map(move |((key, hash), row)| ((key, hash % modulus), row));
+            let (oks, errs) = if validating {
                 let stage = build_topk_stage_inner::<G, Result<Row, Row>>(
-                    collection, order_key, modulus, offset, limit, arity,
+                    &input, order_key, offset, limit, arity,
                 );
+
                 let debug_name = debug_name.to_string();
                 let (oks, errs) =
                     stage.map_fallible("Demuxing Errors", move |((k, h), result)| match result {
@@ -284,12 +288,16 @@ where
                 (oks, Some(errs))
             } else {
                 (
-                    build_topk_stage_inner::<G, Row>(
-                        collection, order_key, modulus, offset, limit, arity,
-                    ),
+                    build_topk_stage_inner::<G, Row>(&input, order_key, offset, limit, arity),
                     None,
                 )
-            }
+            };
+            (
+                oks.negate()
+                    .concat(&input)
+                    .consolidate_named::<RowKeySpine<_, _, _>>("Consolidated TopK"),
+                errs,
+            )
         }
 
         // To provide a robust incremental orderby-limit experience, we want to avoid grouping
@@ -301,9 +309,8 @@ where
         // updates, and to reduce the amount of work on the critical path for updates. The cost is
         // a larger number of arrangements when this optimization does nothing beneficial.
         fn build_topk_stage_inner<G, R>(
-            collection: Collection<G, ((Row, u64), Row), Diff>,
+            input: &Collection<G, ((Row, u64), Row), Diff>,
             order_key: Vec<mz_expr::ColumnOrder>,
-            modulus: u64,
             offset: usize,
             limit: Option<usize>,
             arity: usize,
@@ -313,10 +320,9 @@ where
             G::Timestamp: Lattice,
             R: MaybeValidatingRow<Row, Row>,
         {
-            let input = collection.map(move |((key, hash), row)| ((key, hash % modulus), row));
             // We only want to arrange parts of the input that are not part of the actual output
             // such that `input.concat(&negated_output.negate())` yields the correct TopK
-            let negated_output = input
+            input
                 .arrange_named::<RowSpine<(Row, u64), _, _, _>>("Arranged TopK input")
                 .reduce_abelian::<_, RowSpine<_, _, _, _>>("Reduced TopK input", {
                     move |_key, source, target: &mut Vec<(R, Diff)>| {
@@ -373,37 +379,32 @@ where
                         // the `offset` and `limit` constraints.
                         for index in indexes.into_iter() {
                             let (row, mut diff) = source[index];
+                            if !diff.is_positive() {
+                                continue;
+                            }
+                            // If we are still skipping early records ...
+                            if offset > 0 {
+                                let to_skip = std::cmp::min(offset, usize::try_from(diff).unwrap());
+                                offset -= to_skip;
+                                diff -= Diff::try_from(to_skip).unwrap();
+                            }
+                            // We should produce at most `limit` records.
+                            // TODO(benesch): avoid dangerous `as` conversion.
+                            #[allow(clippy::as_conversions)]
+                            if let Some(limit) = &mut limit {
+                                diff = std::cmp::min(diff, Diff::try_from(*limit).unwrap());
+                                *limit -= diff as usize;
+                            }
+                            // Output the indicated number of rows.
                             if diff > 0 {
-                                // If we are still skipping early records ...
-                                if offset > 0 {
-                                    let to_skip =
-                                        std::cmp::min(offset, usize::try_from(diff).unwrap());
-                                    offset -= to_skip;
-                                    diff -= Diff::try_from(to_skip).unwrap();
-                                }
-                                // We should produce at most `limit` records.
-                                // TODO(benesch): avoid dangerous `as` conversion.
-                                #[allow(clippy::as_conversions)]
-                                if let Some(limit) = &mut limit {
-                                    diff = std::cmp::min(diff, Diff::try_from(*limit).unwrap());
-                                    *limit -= diff as usize;
-                                }
-                                // Output the indicated number of rows.
-                                if diff > 0 {
-                                    // Emit retractions for the elements actually part of
-                                    // the set of TopK elements.
-                                    target.push((R::ok(row.clone()), -diff));
-                                }
+                                // Emit retractions for the elements actually part of
+                                // the set of TopK elements.
+                                target.push((R::ok(row.clone()), -diff));
                             }
                         }
                     }
                 })
-                .as_collection(|k, v| (k.clone(), v.clone()));
-
-            negated_output
-                .negate()
-                .concat(&input.map(move |((k, h), r)| ((k, h), R::ok(r))))
-                .consolidate_named::<RowKeySpine<_, _, _>>("Consolidated TopK")
+                .as_collection(|k, v| (k.clone(), v.clone()))
         }
 
         fn render_top1_monotonic<G>(
```

File: src/compute/src/render/reduce.rs
Comment (tokenrove) [reply src/compute/src/render/reduce.rs:779]: Oops, thanks.  I really need to setup here the pre-push hook I usually use, which restricts pushing the patterns XXX and WIP, which is exactly why I litter my working code with them...

**Comment by vmarcos** [comment]
> > Post-review comment: Should we also add a test based on the scenario in [#17963 (comment)](https://github.com/MaterializeInc/materialize/issues/17963#issuecomment-1477630193) to confirm that we produce errors adequately? This can be done as an additional `mzcompose` workflow in `test/cluster`.
> 
> Yes, I think each of these kind of similar issues should all have tests, since they're easy to reproduce in unsafe mode. I was thinking that I would do these as testdrive tests, just assuring we get the user-level error back, since we already have a test that assures errors go into the logs, and testdrive should be a little lighter-weight. (Though perhaps testdrive should also grow a function to inspect logs.)

Indeed, I've been using `mzcompose` workflows exactly to check the logs as part of the test. But I believe that we hit enough of a pattern and enough of checking there to be satisfied with only looking at the user-level errors in a testdrive test here.

File: src/compute/src/render/mod.rs
Comment (tokenrove) [reply src/compute/src/render/mod.rs]: This works nicely in fact.  Great idea.  I would like to think a little bit about the name, since this signals both that we're validating, and gives us the error constructor.

**Comment by tokenrove** [comment]
Thanks for all the careful reviews.  If I could get any opinions on the naming of `into_error`, that seems to be the last detail before I can merge this.  (It has the merit that it's an idiomatic Rust expression; does it adequately convey in the code that uses it that validation happens as a consequence of it returning Some?)

File: src/compute/src/render/top_k.rs
Comment (vmarcos) [reply src/compute/src/render/top_k.rs]: Ah, and by the way, the same pattern applies also to the min/max aggregate variant!

Comment (tokenrove) [reply src/compute/src/render/top_k.rs]: Oof, thanks, I forgot.

**Pull request merged**
